### PR TITLE
[docs] Add a guide on how to generate a blurhash on a server

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Closes ENG-7096.
Adds a guide on how to generate a blurhash on a server

# Test Plan

image

### CODE REVIEW

File: docs/pages/versions/unversioned/sdk/image.mdx [hunk 1]
```diff
@@ -74,3 +75,59 @@ import { Image } from 'expo-image';
...
+// Sharp allows you to recieve a data buffer from the uploaded image.
+const sharp = require('sharp');
+// Import the encode function from the blurhash package.
+const { encode } = require('blurhash');
+
+// Initilize `multer`.
+const upload = multer();
+```
+
+Assuming app is a variable that holds our express server, we can create an endpoint that accepts an image and returns a JSON response containing the generated blurhash.
```
Comment (tsapeta) [inline docs/pages/versions/unversioned/sdk/image.mdx:103]:
```suggestion
Assuming the `app` is a variable that holds our Express server, we can create an endpoint that accepts an image and returns a JSON response containing the generated blurhash.
```

File: docs/pages/versions/unversioned/sdk/image.mdx [hunk 2]
```diff
@@ -74,3 +75,59 @@ import { Image } from 'expo-image';
...
+  const { file } = req;
+  // If the file is not available we're returning with error.
+  if (file === null) {
+    res.status(400).json({ message: 'Image is missing' });
+    return;
+  }
+
+  // Users can specify number of components in each axies.
+  const compoenentX = req.body.componentX ?? 4;
+  const compoenentY = req.body.componentY ?? 3;
```
Comment (tsapeta) [inline docs/pages/versions/unversioned/sdk/image.mdx:116]:
```suggestion
  const componentX = req.body.componentX ?? 4;
  const componentY = req.body.componentY ?? 3;
```

File: docs/pages/versions/unversioned/sdk/image.mdx [hunk 3]
```diff
@@ -74,3 +75,59 @@ import { Image } from 'expo-image';
...
+
+  // We're converting provided image to a byte buffer.
+  // Sharp currencurrentlylty supported multiple common formats like JPEG, PNG, WebP, GIF, and AVIF.
```
Comment (tsapeta) [inline docs/pages/versions/unversioned/sdk/image.mdx:119]: > currencurrentlylty

What  ⁉️ 😂

File: docs/pages/versions/unversioned/sdk/image.mdx [hunk 4]
```diff
@@ -74,3 +75,59 @@ import { Image } from 'expo-image';
...
+  const { data, info } = await sharp(file.buffer).ensureAlpha().raw().toBuffer({
+    resolveWithObject: true,
+  });
+
+  const blurhash = encode(data, info.width, info.height, compoenentX, compoenentY);
```
Comment (tsapeta) [inline docs/pages/versions/unversioned/sdk/image.mdx:124]:
```suggestion
  const blurhash = encode(data, info.width, info.height, componentX, componentY);
```

File: docs/pages/versions/unversioned/sdk/image.mdx [hunk 5]
```diff
@@ -74,3 +75,59 @@ import { Image } from 'expo-image';
 ```
 
 <APISection packageName="expo-image" apiName="Image" />
+
+## Generating a blurhash on a server
+
+Images can significantly improve the visual experience, but they can also slow down app/page loading times due to their large file sizes. To overcome this, we can create placeholder images using blurhash algorithm that can provide an immersive experience while deferring the loading of the actual picture until later.
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/image.mdx:81]:
```suggestion
Images can significantly improve the visual experience, however, they can also slow down app/page loading times due to their large file sizes. To overcome this, you can create placeholder images using blurhash algorithm that provides an immersive experience while deferring the loading of the actual picture until later.
```

File: docs/pages/versions/unversioned/sdk/image.mdx [hunk 6]
```diff
@@ -74,3 +75,59 @@ import { Image } from 'expo-image';
...
+
+This brief guide will demonstrate how to create a blurhash of an uploaded image on the backend using `JavaScript` and `Express.js`. However, the same techniques and principles apply to other languages and server technologies.
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/image.mdx:83]:
```suggestion
This guide demonstrates how to create a blurhash of an uploaded image on the backend using JavaScript and Express.js. The same techniques and principles apply to other languages and server technologies.
```

File: docs/pages/versions/unversioned/sdk/image.mdx [hunk 7]
```diff
@@ -74,3 +75,59 @@ import { Image } from 'expo-image';
...
+
+To simplify the process, we will install a few dependencies: [`multer`](https://github.com/expressjs/multer) for handling multipart requests, [`sharp`](https://github.com/lovell/sharp) for converting files to a data buffer, and the official [`blurhash JavaScript package`](https://github.com/woltapp/blurhash/tree/master/TypeScript).
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/image.mdx:85]:
```suggestion
Start by installing a few dependencies: [`multer`](https://github.com/expressjs/multer) for handling multipart requests, [`sharp`](https://github.com/lovell/sharp) for converting files to a data buffer, and the official [`blurhash` JavaScript package](https://github.com/woltapp/blurhash/tree/master/TypeScript).
```

File: docs/pages/versions/unversioned/sdk/image.mdx [hunk 8]
```diff
@@ -74,3 +75,59 @@ import { Image } from 'expo-image';
...
+
+<Terminal cmd={['$ npm install multer sharp blurhash']} />
+
+Next, we need to import all needed functions from installed packages and initialized `multer`:
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/image.mdx:89]:
```suggestion
Next, import all required functions from installed packages and initialize `multer`:
```

File: docs/pages/versions/unversioned/sdk/image.mdx (hunk 1 above)
Comment (amandeepmittal) [reply docs/pages/versions/unversioned/sdk/image.mdx:103]: Adding to the above suggestion, I'd avoid using second person voice (We, Our, etc.)

File: docs/pages/versions/unversioned/sdk/image.mdx [hunk 9]
```diff
@@ -74,3 +75,59 @@ import { Image } from 'expo-image';
...
+  const { data, info } = await sharp(file.buffer).ensureAlpha().raw().toBuffer({
+    resolveWithObject: true,
+  });
+
+  const blurhash = encode(data, info.width, info.height, compoenentX, compoenentY);
+  res.json({ blurhash });
+});
+```
+
+Additionally, the request can include two parameters: `componentX` and `componentY` which are passed through the algorithm. These values can be calculated or hard-coded on the server, or specified by the user. However, they must be within the range of 1 to 9 and have an aspect ratio similar to the uploaded image. A value of 9 will give the best results but may take longer to generate the hash.
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/image.mdx:129]:
```suggestion
Additionally, the request can include two parameters: `componentX` and `componentY`, passed through the algorithm. These values can be calculated or hard-coded on the server or specified by the user. However, they must be within the range of 1 to 9 and have an aspect ratio similar to the uploaded image. A value of 9 will give the best results but may take longer to generate the hash.
```

File: docs/pages/versions/unversioned/sdk/image.mdx [hunk 10]
```diff
@@ -74,3 +75,59 @@ import { Image } from 'expo-image';
...
+
+The process of generating a blurhash can be accomplished in various languages and server technologies, similar to the one using JavaScript. The key step is to locate an encoder for your chosen language, which can often be found in the [`woltapp/blurhash`](https://github.com/woltapp/blurhash#implementations) repository. Once you have the encoder, you will need to obtain a representation of the image. Some libraries use a default image class (for example, the Swift implementation uses `UIImage`), but in other cases, you will have to provide raw byte data. Therefore, be sure to check the encoder readme to confirm the expected data format.
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/image.mdx:131]:
```suggestion
The process of generating a blurhash can be accomplished in various languages and server technologies, similar to the one using JavaScript. The key step is to locate an encoder for your chosen language, which can often be found in the [`woltapp/blurhash`](https://github.com/woltapp/blurhash#implementations) repository. Once you have the encoder, you will need to obtain a representation of the image. Some libraries use a default image class (for example, the Swift implementation uses `UIImage`). In other cases, you will have to provide raw byte data. Make sure to check the encoder's documentation to confirm the expected data format.
```

File: docs/pages/versions/unversioned/sdk/image.mdx [hunk 11]
```diff
@@ -74,3 +75,59 @@ import { Image } from 'expo-image';
...
+
+> Note: when working with raw byte data, it is crucial to ensure that all pixels are represented by 4 numbers - the alpha layer is present. Failing to do so may lead to errors like "width and height must match the pixels array".
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/image.mdx:133]:
```suggestion
> When working with raw byte data, it is crucial to ensure that all pixels are represented by 4 numbers - the alpha layer is present. Failing to do so may lead to errors such as "width and height must match the pixels array".
```

This phrase: "the alpha layer is present", doesn't make sense to me. Is the pixel composition represented by 4 numbers is called "the alpha layer"? If yes, we might want to rephrase the sentence and/or add a link, if possible, from the blurhash docs.

**Review by amandeepmittal — COMMENTED** [review]
Thanks for writing this guide! Left some suggestions.

I would like to improve the organization of this guide. I think that guides should be placed either in Assorted Guides or Integrations section (for now). 

I also realize this doc is unreleased, so that can be done later (I'll create a follow up task and once this doc is released, or close to getting released in the next SDK, and we can move the guide to its own separate page.)

File: docs/pages/versions/unversioned/sdk/image.mdx [hunk 12]
```diff
@@ -74,3 +75,59 @@ import { Image } from 'expo-image';
...
+```js
+app.post('/blurhash', upload.single('image'), async (req, res) => {
+  const { file } = req;
+  // If the file is not available we're returning with error.
+  if (file === null) {
+    res.status(400).json({ message: 'Image is missing' });
+    return;
+  }
+
+  // Users can specify number of components in each axies.
```
Comment (Simek) [inline docs/pages/versions/unversioned/sdk/image.mdx:114]:
```suggestion
  // Users can specify number of components in each axes.
```

"Axes" or "Axiis" is a plural - https://en.wiktionary.org/wiki/axis

**Comment by lukmccall** [comment]
> I would like to improve the organization of this guide. I think that guides should be placed either in Assorted Guides or Integrations section (for now).
> 
> I also realize this doc is unreleased, so that can be done later (I'll create a follow up task and once this doc is released, or close to getting released in the next SDK, and we can move the guide to its own separate page.)

We can move it to a different place, but until the `expo-image` wasn't published within Expo go I would instead leave it in the unreleased docs.

File: docs/pages/versions/unversioned/sdk/image.mdx [hunk 13]
```diff
@@ -74,3 +75,59 @@ import { Image } from 'expo-image';
...
+    res.status(400).json({ message: 'Image is missing' });
+    return;
+  }
+
+  // Users can specify number of components in each axes.
+  const componentX = req.body.componentX ?? 4;
+  const componentY = req.body.componentY ?? 3;
+
+  // We're converting provided image to a byte buffer.
+  // Sharp currently supported multiple common formats like JPEG, PNG, WebP, GIF, and AVIF.
```
Comment (Simek) [inline docs/pages/versions/unversioned/sdk/image.mdx:119]:
```suggestion
  // Sharp currently supports multiple common formats like JPEG, PNG, WebP, GIF, and AVIF.
```

Wording nit.

**Comment by amandeepmittal** [comment]
Thanks for the input @lukmccall! I'll follow-up with moving the guide when the `expo-image` is published.

File: docs/pages/versions/unversioned/sdk/image.mdx [hunk 14]
```diff
@@ -74,3 +75,59 @@ import { Image } from 'expo-image';
...
+
+The process of generating a blurhash can be accomplished in various languages and server technologies, similar to the one using JavaScript. The key step is to locate an encoder for your chosen language, which can often be found in the [`woltapp/blurhash`](https://github.com/woltapp/blurhash#implementations) repository. Once you have the encoder, you will need to obtain a representation of the image. Some libraries use a default image class (for example, the Swift implementation uses `UIImage`). In other cases, you will have to provide raw byte data. Make sure to check the encoder's documentation to confirm the expected data format.
+
+> When working with raw byte data, please ensure the alpha layer is present (each pixel is represented by red, green, blue and alpha values). Failing to do so may lead to errors like "width and height must match the pixels array".
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/image.mdx:133]:
```suggestion
> When working with raw byte data, ensure that the alpha layer is present (each pixel is represented by red, green, blue, and alpha values). Failing to do so will lead to errors such as "width and height must match the pixels array".
```

**Review by amandeepmittal — APPROVED** [review]
Lgtm!

One small suggestion for being consistent with verbiage and avoiding phrases such as `please` 😄

**Pull request merged**